### PR TITLE
Hires rho

### PR DESCRIPTION
**Pull request by catherinewatkinson**
This implements the option to perform the particle moving for the density field (be it with ZA or 2LPT) at high resolution. It was done on a backdated version of the code, so some work is needed to get it inline with the latest version of the code. 

I have added some basic tests to Tests/ and there are also some plots in Tests/DensityFieldTestData/ that show the difference in performing the density field particle movement at low and high resolution.

### CODE REVIEW

**Comment by steven-murray** [comment]
Thanks @catherinewatkinson. I'm happy to help with getting the code working with the newest version, if others (eg. @BradGreig and @andreimesinger) can look at some of the plots/tests to make sure it's doing what they think it should be doing.

**Comment by BradGreig** [comment]
@catherinewatkinson @steven-murray I've had a cursory glance of the code and the plots. The behaviour of the plots makes sense and is inline with discussions we've had in the past. I assume it is power rather than dimensionless power in the power spectrum plot?

As for the code (which I appreciate is out-dated and incomplete) I think there could be some further changes. I too would be happy to oversee the integration of this code into the updated version. Some changes are:
1. Renaming the flag (can't think of a good one yet though...)
2. Having both low-res and high-res smoothing options within the code (to enable backwards compatibility)
3. Refactoring the code. Currently there is a considerable amount of redundant code by having one if statement on the outermost level. I think it should be able to be re-written to minimise redundant code.

As I said, I am happy to do all these myself. Just let me know

**Comment by steven-murray** [comment]
@BradGreig awesome, thanks for taking a look. I think this can wait til v3.1, as it shouldn't make any breaking API changes. I'm obviously happy for you to work on it before that though :-) Probably higher priority right now to get the full functionality of previous 21cmFAST though.

**Comment by catherinewatkinson** [comment]
Hi Brad,

All sounds good, it currently has the functionality of being able to run in low res or high res mode, but I definitely agree it could be done in a more elegant fashion to reduce the redundancy of code. There may be ways to speed up the high res version as implemented here.

Cat

**Comment by catherinewatkinson** [comment]
Regarding the plot question, looking at the test unit, I think this should be dimensionless power (i.e. P*k**3/(2*pi**2). However, it's possible I'm mistaken and there is an inconsistency; I did this all pretty quickly since I knew it would ultimately get altered drastically and I knew Andrei was keen to get it, I think for the minihalos stuff.

**Comment by steven-murray** [comment]
Is the code to make the plots in that Test/ directory?

**Comment by catherinewatkinson** [comment]
Yes. In the initial conditions .py test script (I know it's technically also perturb, but I was feeling lazy). I think I might've made the files in a separate python script, but I doubt i got very creative and would've just copied it across when I made the tests. I also think from the fact it's not failing the tests that it must be the same.

**Comment by steven-murray** [comment]
I *might* get time to have a look through this today in a bit more detail.
However, I think at this point that this will be going into v3.1, and it
might be best if @BradGreig gets a chance to go through it first to iron
things out on the C side.

**Comment by andreimesinger** [comment]
This should be pulled in before any Lyman alpha forest work, since the density and velocity fields are better resolved.

**Comment by BradGreig** [comment]
Ok @andreimesinger, I'll try and get to this as soon as I can. I'll work on pulling in the multi-threading branch first though.

**Comment by BradGreig** [comment]
Ok, question for @andreimesinger @catherinewatkinson. I have gone through and added in this high-res implementation of the Zel'dovich/2LPT for the perturbed fields (currently, it is only stored locally). I have a question regarding the down-sampling of this high-res perturbed density field to the low-res density field.

In the first figure, I am showing the density PDF for the original 21cmFAST implementation (perturbed on low-res grid) compared to @catherinewatkinson's implementation (blue) and an alternative (red). Here, the spikiness in the original implementation is gone in the updated implementation.

image

Now, the difference between the blue and red curves is in how the high-res grid is down-sampled to the low-res grid. In the past, when down-sampling, the field in question is filtered by a filter and then sampled in HII_DIM. Now, @catherinewatkinson, instead your implementation is to effectively  average the mass over the entire lower resolution cell volume. In the blue curve, I retain the original 21cmFAST method, down-sampling on the perturbed field (where perturbing was done at high-res).

There is no major difference in the density PDF from either method, both remove the spikiness, which is the desired result from doing the perturbing in the high-res box.

However, there is a considerable difference in the resultant density power spectrum, shown below:
image

In @catherinewatkinson's approach, in averaging the mass over the low-res voxel it smooths the density field, resulting in the turn-over of the density power spectrum. In only using the sampled field (i.e. original approach) the power spectrum does not turn-over (red curve).

My question is, which approach is more appropriate to use? I think the sampling approach (red curve) is more appropriate as having such a large drop-off in the power doesn't seem to be the desired behaviour. The density PS should follow something more closely mimicking the red curve if its goal is to include (quasi) non-linear structure.

**Comment by catherinewatkinson** [comment]
I had a quick skim through the literature and I would say your sampling method looks more consistent. There is a turnover, which is why on my quick check I wasn't alarmed by it, but that is seen in P(k) not Delta(k) as plotted here. I didn't look at any of this stuff in great detail beyond simply ensuring the PDF looked okay. So it's very likely the implementation I did needs refining as suggested.

**Comment by andreimesinger** [comment]
hi brad, if i am understanding correctly, the difference is in the downsampling, with one method doing it in k space with a spherical top hat and the other in real space with a box car...

in general, downsampling in k space is preferred over box car, though one is still prone to aliasing.  indeed the matter power spectrum should not be turning over like that, even though part of that rise might still be due tk aliasing, i think that is preferred

thanks
-a

**Comment by BradGreig** [comment]
@andreimesinger both are smoothed with a real-space filter (as was always the case with 21cmFAST), the difference is the sampling of the filtered hi-res box.

I have,
```
for (i=0; i<user_params->HII_DIM; i++){
    for (j=0; j<user_params->HII_DIM; j++){
        for (k=0; k<user_params->HII_DIM; k++){
            *((float *)LOWRES_density_perturb + HII_R_FFT_INDEX(i,j,k) 
            =
            (*((float *)HIRES_density_perturb + R_FFT_INDEX(i*f_pixel_factor,j*f_pixel_factor,k*f_pixel_factor))/(float)TOT_NUM_PIXELS);
        }
    }
}
```
whereas @catherinewatkinson approach is to average all high resolution elements into a low-res voxel. Hence the power spectrum shape.

```
for (i=0; i<user_params->DIM; i++){
    for (j=0; j<user_params->DIM; j++){
        for (k=0; k<user_params->DIM; k++){
            *((float *)LOWRES_density_perturb + HII_R_FFT_INDEX((i + 0.)/f_pixel_factor),(j + 0.)/f_pixel_factor,(k + 0.)/f_pixel_factor) 
            +=
            (*((float *)HIRES_density_perturb + R_FFT_INDEX(i,j,k))/(float)TOT_NUM_PIXELS) - 1. )/mass_factor;
        }
    }
}
```
In any case, I'll go with what I have implemented as that seems to be the preferred.

**Comment by BradGreig** [comment]
Hi @steven-murray, can you have a look at why this is now failing. One test fails, which I don't understand why given I updated the data for it. But it doesn't output anything useful because the tests never finish as there seems to be issues with travis-ci.

Turning off all integration tests still fails, same issue with travis-ci. Running things locally passes without any problems. I am running with pre-commit and everything.

**Comment by steven-murray** [comment]
Huh. Very strange errors. I'm encountering errors on #127 as well (with the integration tests). I'll plug away at it today (I have it running through valgrind currently). I've found a way to reduce the test run time fairly significantly (if everything works), so that will be nice.

**Comment by BradGreig** [comment]
@steven-murray @andreimesinger @catherinewatkinson, I have marked this PR for review. If you get the time can you have a quick look through. Its failing its test on travis, which @steven-murray is looking into. However, the cause should be due to anything on the C side of things.

File: src/py21cmfast/outputs.py [hunk 1]
```diff
@@ -104,9 +109,21 @@ def _init_arrays(self):
         self.lowres_vz_2LPT = np.zeros(
             self.user_params.HII_tot_num_pixels, dtype=np.float32
         )
+
```
Comment (steven-murray) [inline src/py21cmfast/outputs.py:112]: Should we be initializing these _only_ if doing the `PERTURB_ON_HIGH_RES`?

Comment (BradGreig) [reply src/py21cmfast/outputs.py:112]: There are quite a few different arrays (in most classes) that should only be initialised under certain parameter combination. Issues #27, #63 and #69 discuss this to various degrees. At this point I am just aiming to get in all pieces of functionality, before going through and initialising only specific pieces.

Comment (steven-murray) [reply src/py21cmfast/outputs.py:112]: Yup, fair enough. We can address this when addressing those issues.

**Comment by BradGreig** [comment]
Hey @steven-murray, I have finally manage to get all the integration tests to pass. However, one test seems to keep failing, `test_readability` from `tests/test_output_structs.py`. Given you added this test, I figure you are the best person to resolve the issue.

This has actually been an issue in this branch for quite some time.

**Comment by steven-murray** [comment]
Awesome @BradGreig ! I'll go and fix that. It's probably something that broke on rebase.

**Comment by steven-murray** [comment]
OK I have gotten this to all work. The problem was actually the memory usage -- there were a couple of functions that were using the default user params and blowing the memory up.

**Pull request merged**